### ISSUE TITLE
Mario & Luigi Paper Jam - Uniform block size did not match!

### ISSUE THREAD

**Issue by pippo2931** [citra-emu/citra#2213]
Mario & Luigi Paper Jam crashes when loading a save game. OS X latest local build.

Driver is the vanilla driver of macOS Sierra (10.12.1) for the Intel Iris Pro.
Yes, size is zero and the shader is failing.

```
^[[0m^[[1;35m[  16.574664] Render.OpenGL <Critical> video_core/renderer_opengl/gl_rasterizer_cache.cpp:GetFramebufferSurfaces:556: Color and depth framebuffer memory regions overlap; overlapping framebuffers not supported!
^[[0m^[[1;35m[  16.574687] Render.OpenGL <Critical> video_core/renderer_opengl/gl_rasterizer_cache.cpp:GetFramebufferSurfaces:556: Color and depth framebuffer memory regions overlap; overlapping framebuffers not supported!
^[[0m^[[1;35m[  16.574707] Render.OpenGL <Critical> video_core/renderer_opengl/gl_rasterizer_cache.cpp:GetFramebufferSurfaces:556: Color and depth framebuffer memory regions overlap; overlapping framebuffers not supported!
^[[0m^[[1;35m[  16.755412] Render.OpenGL <Critical> video_core/renderer_opengl/gl_shader_gen.cpp:AppendSource:80: Unknown source op 6
^[[0m^[[1;31m[  16.760042] Render.OpenGL <Error> video_core/renderer_opengl/gl_shader_util.cpp:LoadProgram:58: Error compiling fragment shader:
ERROR: 0:63: Index 3 beyond bounds (size 3)
ERROR: 0:63: Index 3 beyond bounds (size 3)
ERROR: 0:64: Use of undeclared identifier 'surface_normal'
ERROR: 0:69: Use of undeclared identifier 'normal'
ERROR: 0:70: Use of undeclared identifier 'normal'

^[[0m^[[1;31m[  16.760082] Render.OpenGL <Error> video_core/renderer_opengl/gl_shader_util.cpp:LoadProgram:81: Error linking shader:
ERROR: One or more attached shaders not successfully compiled

^[[0m^[[1;35m[  16.760095] Debug <Critical> video_core/renderer_opengl/gl_rasterizer.cpp:operator():1073: Assertion Failed!
Uniform block size did not match!
 ```

*Posts re-arranged and re-formated by JayFoxRox*

**Comment by wwylele** [comment]
cc @JayFoxRox

**Comment by JayFoxRox** [comment]
Which GPU driver? We had some problems like that in the past and they turned out to be some kind of undefined behaviour / bad wording in the GL spec which was related to vector alignment at the end of the buffer.

To avoid such errors, I re-arranged the uniform block in https://github.com/citra-emu/citra/commit/c900c092e329d2e78b9663c1c3fe401da7faaea5 .
So it would be weird if that particular commit (or the one just after it) breaks compilation this way.
(This got a lot of testing and has been working fine for months on pretty much any GPU driver)

Additionally this error comes up if some shader compilation failed (which would return a block size of 0).
However, we didn't get any issue in the past for that either.

I'm assuming the real error is elsewhere or it's a broken GPU driver / undefined behaviour in the spec.

What GPU / GPU driver are you using? Which version of the OS? What's the log of the GL shader compilation (Not sure if that is part of the citra-log?)? What's in the citra-log?
Also, please always mention the current commit in issues because otherwise we'll have to look up the commit by the GitHub issue timestamp which is kind of stupid.

(I've removed my e-mail address from the post above. I'm aware it's part of the commit, but I feel like it gets far more exposure here than it should)

**Comment by pippo2931** [comment]
Driver is the vanilla driver of macOS Sierra (10.12.1) for the Intel Iris Pro.
Yes, size is zero and the shader is failing.

^[[0m^[[1;35m[  16.574664] Render.OpenGL <Critical> video_core/renderer_opengl/gl_rasterizer_cache.cpp:GetFramebufferSurfaces:556: Color and depth framebuffer memory regions overlap; overlapping framebuffers not supported!
^[[0m^[[1;35m[  16.574687] Render.OpenGL <Critical> video_core/renderer_opengl/gl_rasterizer_cache.cpp:GetFramebufferSurfaces:556: Color and depth framebuffer memory regions overlap; overlapping framebuffers not supported!
^[[0m^[[1;35m[  16.574707] Render.OpenGL <Critical> video_core/renderer_opengl/gl_rasterizer_cache.cpp:GetFramebufferSurfaces:556: Color and depth framebuffer memory regions overlap; overlapping framebuffers not supported!
^[[0m^[[1;35m[  16.755412] Render.OpenGL <Critical> video_core/renderer_opengl/gl_shader_gen.cpp:AppendSource:80: Unknown source op 6
^[[0m^[[1;31m[  16.760042] Render.OpenGL <Error> video_core/renderer_opengl/gl_shader_util.cpp:LoadProgram:58: Error compiling fragment shader:
ERROR: 0:63: Index 3 beyond bounds (size 3)
ERROR: 0:63: Index 3 beyond bounds (size 3)
ERROR: 0:64: Use of undeclared identifier 'surface_normal'
ERROR: 0:69: Use of undeclared identifier 'normal'
ERROR: 0:70: Use of undeclared identifier 'normal'

^[[0m^[[1;31m[  16.760082] Render.OpenGL <Error> video_core/renderer_opengl/gl_shader_util.cpp:LoadProgram:81: Error linking shader:
ERROR: One or more attached shaders not successfully compiled

^[[0m^[[1;35m[  16.760095] Debug <Critical> video_core/renderer_opengl/gl_rasterizer.cpp:operator():1073: Assertion Failed!
Uniform block size did not match!

**Comment by JayFoxRox** [comment]
Can you please add the full shader source? (can be dumped using apitrace or by stepping through the code I guess?)

So far this doesn't look like a regression in the specified piece of code / commit at all. It might be related to JIT / non-JIT too?

**Comment by pippo2931** [comment]
The trace file is big, I compressed the dump file.
[citra-qt.dump.gz](https://github.com/citra-emu/citra/files/614604/citra-qt.dump.gz)

**Comment by JayFoxRox** [comment]
The gist of it is this
```
447909 glCreateShader(type = GL_VERTEX_SHADER) = 619
447910 glCreateShader(type = GL_FRAGMENT_SHADER) = 620
447911 glShaderSource(shader = 619, count = 1, string = &"
```
```
#version 330 core
layout(location = 0) in vec4 vert_position;
layout(location = 1) in vec4 vert_color;
layout(location = 2) in vec2 vert_texcoord0;
layout(location = 3) in vec2 vert_texcoord1;
layout(location = 4) in vec2 vert_texcoord2;
layout(location = 5) in float vert_texcoord0_w;
layout(location = 6) in vec4 vert_normquat;
layout(location = 7) in vec3 vert_view;

out vec4 primary_color;
out vec2 texcoord[3];
out float texcoord0_w;
out vec4 normquat;
out vec3 view;

void main() {
    primary_color = vert_color;
    texcoord[0] = vert_texcoord0;
    texcoord[1] = vert_texcoord1;
    texcoord[2] = vert_texcoord2;
    texcoord0_w = vert_texcoord0_w;
    normquat = vert_normquat;
    view = vert_view;
    gl_Position = vec4(vert_position.x, vert_position.y, -vert_position.z, vert_position.w);
}
```
```
", length = NULL)
447912 glCompileShader(shader = 619)
447913 glGetShaderiv(shader = 619, pname = GL_COMPILE_STATUS, params = &1)
447914 glGetShaderiv(shader = 619, pname = GL_INFO_LOG_LENGTH, params = &0)
447915 glShaderSource(shader = 620, count = 1, string = &"
```
```

#version 330 core
#define NUM_TEV_STAGES 6
#define NUM_LIGHTS 8
#define LIGHTING_LUT_SIZE 256
#define FLOAT_255 (255.0 / 256.0)

in vec4 primary_color;
in vec2 texcoord[3];
in float texcoord0_w;
in vec4 normquat;
in vec3 view;

in vec4 gl_FragCoord;

out vec4 color;

struct LightSrc {
    vec3 specular_0;
    vec3 specular_1;
    vec3 diffuse;
    vec3 ambient;
    vec3 position;
    float dist_atten_bias;
    float dist_atten_scale;
};

layout (std140) uniform shader_data {
    vec2 framebuffer_scale;
    int alphatest_ref;
    float depth_scale;
    float depth_offset;
    int scissor_x1;
    int scissor_y1;
    int scissor_x2;
    int scissor_y2;
    vec3 fog_color;
    vec3 lighting_global_ambient;
    LightSrc light_src[NUM_LIGHTS];
    vec4 const_color[NUM_TEV_STAGES];
    vec4 tev_combiner_buffer_color;
};

uniform sampler2D tex[3];
uniform sampler1D lut[6];
uniform usampler1D fog_lut;

// Rotate the vector v by the quaternion q
vec3 quaternion_rotate(vec4 q, vec3 v) {
    return v + 2.0 * cross(q.xyz, cross(q.xyz, v) + q.w * v);
}

void main() {
vec4 primary_fragment_color = vec4(0.0);
vec4 secondary_fragment_color = vec4(0.0);
if (!(gl_FragCoord.x >= scissor_x1 && gl_FragCoord.y >= scissor_y1 && gl_FragCoord.x < scissor_x2 && gl_FragCoord.y < scissor_y2)) discard;
float z_over_w = 1.0 - gl_FragCoord.z * 2.0;
float depth = z_over_w * depth_scale + depth_offset;
vec4 diffuse_sum = vec4(0.0, 0.0, 0.0, 1.0);
vec4 specular_sum = vec4(0.0, 0.0, 0.0, 1.0);
vec3 light_vector = vec3(0.0);
vec3 refl_value = vec3(0.0);
vec3 surface_normal = 2.0 * texture(tex[3], texcoord[3]).rgb - 1.0;
vec3 normal = normalize(quaternion_rotate(normquat, surface_normal));
light_vector = normalize(light_src[0].position);
refl_value.r = 1.0;
refl_value.g = refl_value.r;
refl_value.b = refl_value.r;
diffuse_sum.rgb += ((light_src[0].diffuse * max(dot(light_vector, normal), 0.0)) + light_src[0].ambient) * 1.0;
specular_sum.rgb += (((1.000000 * texture(lut[0], (FLOAT_255 * clamp(max(dot(normal, normalize(view)), 0.f), 0.0, 1.0)))[0]) * light_src[0].specular_0) + (1.0 * refl_value * light_src[0].specular_1)) * 1.0 * 1.0;
diffuse_sum.rgb += lighting_global_ambient;
primary_fragment_color = clamp(diffuse_sum, vec4(0.0), vec4(1.0));
secondary_fragment_color = clamp(specular_sum, vec4(0.0), vec4(1.0));
vec4 combiner_buffer = vec4(0.0);
vec4 next_combiner_buffer = tev_combiner_buffer_color;
vec4 last_tex_env_out = vec4(0.0);
vec3 color_results_0[3] = vec3[3](const_color[0].aaa, const_color[0].rgb, primary_color.rgb);
vec3 color_output_0 = clamp(color_results_0[0] * color_results_0[2] + color_results_0[1] * (vec3(1.0) - color_results_0[2]), vec3(0.0), vec3(1.0));
float alpha_results_0[3] = float[3](primary_color.a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a);
float alpha_output_0 = clamp(alpha_results_0[0], 0.0, 1.0);
last_tex_env_out = vec4(clamp(color_output_0 * 1.0, vec3(0.0), vec3(1.0)),clamp(alpha_output_0 * 1.0, 0.0, 1.0));
combiner_buffer = next_combiner_buffer;
vec3 color_results_1[3] = vec3[3](last_tex_env_out.rgb, const_color[1].rgb, primary_color.aaa);
vec3 color_output_1 = clamp(color_results_1[0] * color_results_1[2] + color_results_1[1] * (vec3(1.0) - color_results_1[2]), vec3(0.0), vec3(1.0));
float alpha_results_1[3] = float[3](last_tex_env_out.a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a);
float alpha_output_1 = clamp(alpha_results_1[0], 0.0, 1.0);
last_tex_env_out = vec4(clamp(color_output_1 * 1.0, vec3(0.0), vec3(1.0)),clamp(alpha_output_1 * 1.0, 0.0, 1.0));
combiner_buffer = next_combiner_buffer;
vec3 color_results_2[3] = vec3[3](last_tex_env_out.rgb, vec4(0.0).rgb, secondary_fragment_color.rgb);
vec3 color_output_2 = clamp(color_results_2[0] * color_results_2[1] + color_results_2[2], vec3(0.0), vec3(1.0));
float alpha_results_2[3] = float[3](last_tex_env_out.a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a);
float alpha_output_2 = clamp(alpha_results_2[0], 0.0, 1.0);
last_tex_env_out = vec4(clamp(color_output_2 * 1.0, vec3(0.0), vec3(1.0)),clamp(alpha_output_2 * 1.0, 0.0, 1.0));
combiner_buffer = next_combiner_buffer;
vec3 color_results_3[3] = vec3[3](const_color[3].rgb, last_tex_env_out.rgb, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).aaa);
vec3 color_output_3 = clamp(color_results_3[0] * color_results_3[2] + color_results_3[1] * (vec3(1.0) - color_results_3[2]), vec3(0.0), vec3(1.0));
float alpha_results_3[3] = float[3](last_tex_env_out.a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a, textureProj(tex[0], vec3(texcoord[0], texcoord0_w)).a);
float alpha_output_3 = clamp(alpha_results_3[0], 0.0, 1.0);
last_tex_env_out = vec4(clamp(color_output_3 * 1.0, vec3(0.0), vec3(1.0)),clamp(alpha_output_3 * 1.0, 0.0, 1.0));
combiner_buffer = next_combiner_buffer;
combiner_buffer = next_combiner_buffer;
combiner_buffer = next_combiner_buffer;
float fog_index = depth * 128.0;
float fog_i = clamp(floor(fog_index), 0.0, 127.0);
float fog_f = fog_index - fog_i;
uint fog_lut_entry = texelFetch(fog_lut, int(fog_i), 0).r;
float fog_lut_entry_difference = float(int((fog_lut_entry & 0x1FFFU) << 19U) >> 19);
float fog_lut_entry_value = float((fog_lut_entry >> 13U) & 0x7FFU);
float fog_factor = (fog_lut_entry_value + fog_lut_entry_difference * fog_f) / 2047.0;
fog_factor = clamp(fog_factor, 0.0, 1.0);
last_tex_env_out.rgb = mix(fog_color.rgb, last_tex_env_out.rgb, fog_factor);
gl_FragDepth = depth;
color = last_tex_env_out;
}
```
```
", length = NULL)
447916 glCompileShader(shader = 620)
447917 glGetShaderiv(shader = 620, pname = GL_COMPILE_STATUS, params = &0)
447918 glGetShaderiv(shader = 620, pname = GL_INFO_LOG_LENGTH, params = &250)
447919 glGetShaderInfoLog(shader = 620, bufSize = 250, length = NULL, infoLog = "ERROR: 0:63: Index 3 beyond bounds (size 3)
ERROR: 0:63: Index 3 beyond bounds (size 3)
ERROR: 0:64: Use of undeclared identifier 'surface_normal'
ERROR: 0:69: Use of undeclared identifier 'normal'
ERROR: 0:70: Use of undeclared identifier 'normal'
")
447920 glCreateProgram() = 621
447921 glAttachShader(program = 621, shader = 619)
447922 glAttachShader(program = 621, shader = 620)
447923 glLinkProgram(program = 621)
447924 glGetProgramiv(program = 621, pname = GL_LINK_STATUS, params = &0)
447925 glGetProgramiv(program = 621, pname = GL_INFO_LOG_LENGTH, params = &63)
447926 glGetProgramInfoLog(program = 621, bufSize = 63, length = NULL, infoLog = "ERROR: One or more attached shaders not successfully compiled
")
```

**Comment by JayFoxRox** [comment]
```
vec3 surface_normal = 2.0 * texture(tex[3], texcoord[3]).rgb - 1.0;
```

Texture unit 3 is the produceral texture unit which is currently not emulated.
This can't be a regression as the texture unit was never emulated.
If this was working fine before it was a bug elsewhere (probably the GPU driver being silent about it). If you can still manage to run it somehow, please tell us which driver + version of Citra.

**Comment by pippo2931** [comment]
Same driver, loading the same save game, used to work in #1701.
I went back to 57a5fc4 and I can see that the shader is different.
[citra-qt2.dump.gz](https://github.com/citra-emu/citra/files/614812/citra-qt2.dump.gz)

**Comment by JayFoxRox** [comment]
Right.. but that also uses a different CPU implementation etc.
That commit.. again.. is not changing anything related to that bug.

Also I'm unable to find the corresponding shader in that apitrace dump. Please tell us which draw call / shader object this shader maps to in the second log.

Lastly, could you please re-test on master without the JIT and also bisect the issue properly?

**Comment by syedraza1997** [comment]
I am facing the same problem

**Comment by syedraza1997** [comment]
Plz tell me how to fix it

**Comment by pippo2931** [comment]
There is no fix,let's hope that somebody that understands shaders will work on it.-G

**Comment by JayFoxRox** [comment]
@pippo2931 you don't really have to understand shaders to fix his - it's a very basic hack which could make this work. The reason this issue is still opened is: We don't know why this happens.

**Comment by syedraza1997** [comment]
When this issue will be fixed

**Comment by syedraza1997** [comment]
Any update about this issue?

**Comment by RavenHome1** [comment]
@syedraza1997 no there does not seem to be any updates regarding this issue

instead of posting question on this issue just add citra to your watch on github and you will be notified if any PR that fixes this issue appears

**Comment by syedraza1997** [comment]
Still not fixed?
I am using windows 10

**Comment by syedraza1997** [comment]
When this issue will be fixed

**Comment by wwylele** [comment]
@syedraza1997 (refer to deleted comments) Please stop spamming here. This issue tracker is for development, not for you to asking questions repeatedly.

**Comment by syedraza1997** [comment]
Look I am also facing this problem since 3 months. Do you have solution for this?

**Comment by wwylele** [comment]
Closed by #2697
